keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,376 @@
+- Makes it complex to know which operator-related features are available in which versions.  Operator revision for different app versions are unrelated
+- Upgrading from `3.11.4-0.1.2` to `3.12.0-0.1.2` might result in missing some operator-related features that were present in `3.11.4-0.1.2`, even if unlikely
+
+### Example hypothetical timeline for Apache Cassandra releases
+
+_Resolved version_ is the combination of _app version_ and _operator version_ used when ordering the package versions.
+
+| Time | Resolved Version | App version | Operator version | KUDO API version | Comment                                    | Operator revision change |
+| ---- | ---------------- | ----------- | ---------------- | ---------------- | ------------------------------------------ | ------------------------ |
+| T0   | 3.11.4-0.1.0     |      3.11.4 |            0.1.0 | v1beta1          | Initial release based on Apache C\* 3.11.x | -                        |
+| T1   | 3.11.4-0.2.0     |      3.11.4 |            0.2.0 | v1beta1          | Important bug fix in operator-related code | Minor bump               |
+| T2   | 3.11.4-0.2.1     |      3.11.4 |            0.2.1 | v1beta1          | Small bug fix in operator-related code     | Patch bump               |
+| T3   | 3.12.0-0.1.0     |      3.12.0 |            0.1.0 | v1beta1          | Apache C\* 3.12.x release                  | Reset (0.1.0)            |
+| T4   | 3.11.4-0.3.0     |      3.11.4 |            0.3.0 | v1beta1          | Operator-related feature A added to 3.11.x | Minor bump               |
+| T4   | 3.12.0-0.2.0     |      3.12.0 |            0.2.0 | v1beta1          | Operator-related feature A added to 3.12.x | Minor bump               |
+| T5   |  4.0.0-0.1.0     |       4.0.0 |            0.1.0 | v1beta1          | Apache C\* 4.0.x release                   | Reset (0.1.0)            |
+| T6   | 3.11.4-0.4.0     |      3.11.4 |            0.4.0 | v1beta1          | Operator-related feature B added to 3.11.x | Minor bump               |
+| T6   | 3.12.0-0.3.0     |      3.12.0 |            0.3.0 | v1beta1          | Operator-related feature B added to 3.12.x | Minor bump               |
+| T6   |  4.0.0-0.2.0     |       4.0.0 |            0.2.0 | v1beta1          | Operator-related feature B added to 4.0.x  | Minor bump               |
+| T7   | 3.11.4-1.0.0     |      3.11.4 |            1.0.0 | v1               | KUDO API version change                    | Major bump               |
+| T7   | 3.12.0-1.0.0     |      3.12.0 |            1.0.0 | v1               | KUDO API version change                    | Major bump               |
+| T7   |  4.0.0-1.0.0     |       4.0.0 |            1.0.0 | v1               | KUDO API version change                    | Major bump               |
+
+### Required and Suggested Changes
+
+1. Make the _app version_ a required field, as it takes precedence over an (optional) operator version
+2. Make the _operator version_ field optional.
```

Comment:
I'm still not fully sold on this. To me it would make more sense to keep the `operatorVersion` required, and `appVersion` optional, and only if it's present let it take precedence. 
- This would make more sense for operators that don't really have a (single) application
- It would keep existing operators compatible. 

On the other hand, implementing it as described here works as well, and it may be just a personal preference thing, so I'm ok with it.

Reply:
An operator bundles an application and ideally is trivial. Hence the requirement for an `appVersion`. Of course, an operator won't rarely be trivial and will have revisions. This can be covered by the now optional `operatorVersion`.
Yes, operators without a single or "main" application aren't covered by this approach.

Reply:
I was also under the impression that the *application version* would be made optional and the *operator version* (or revision) would continue being required. Given that a KUDO Operator can either have an underlying main *application* or not.

Reply:
That's the current status quo: `appVersion` is already optional, `operatorVersion` is required. We decided to swap this due to the assumptions described above.

Reply:
tbh, I have a kind of strong opposition against this.
I may have missed some discussions about the `appVersion` vs. `operatorVersion` optional/required, but I feel I missed the part where we decided that. I remember having some discussions about it, but no decision.

Again: If we do it this way:
1. We do not have support for operators without a single/main application
2. We exchange the required operatorVersion where we have defined semantics and ordering for a string that doesn't provide semVer semantics.
3. We introduce a breaking change

Point 3 is arguably not that big of a deal, but I feel that point 1 and 2 could bring us a lot of pain in the future.

@kensipe @alenkacz @gerred @zmalik @zen-dog Could you please share your opinion on this point?

Reply:
I agree with you @ANeumann82, given that we have https://github.com/kudobuilder/operators/tree/master/repository/flink/docs/demo/financial-fraud (and a recent blog post highlighting this) which is an operator without a main application.

Reply:
Afair from our discussions, introducing breaking changes now is a much smaller concern than leaving things in a way that might be difficult to deal with in the future.

Reply:
Seems that I also missed "the decision". However,
- `operatorVersion` should be required and SemVer.  This is already a case and I don't see why we would relax this. And for edge cases where people don't need it, it stays `0.0.0` forever. Ez.
- `appVersion` *can* be optional to preserve compatibility with existing operators.

Reply:
as we have discussed (and it sounds like we need to meet again)

we need to support:
1. operators with 0 base techs
2. operators with 1 base tech 
3. and operators with many

I thought we agreed that under conditions 1 and 3 that the author of the operator would need to specify a project managed version and that app version was required.  

We also decided at the end of those series of meetings what is written here... that app version is required and operator version (or revision) could be useful and for many operators (including option 2 above) we would use it to distinguish increments in kudo versions.